receiver/receivercreator/hints.go

```diff
@@ -0,0 +1,250 @@
+// Copyright The OpenTelemetry Authors
+// SPDX-License-Identifier: Apache-2.0
+
+package receivercreator // import "github.com/open-telemetry/opentelemetry-collector-contrib/receiver/receivercreator"
+
+import (
+	"fmt"
+
+	"go.uber.org/zap"
+	"gopkg.in/yaml.v3"
+
+	"github.com/open-telemetry/opentelemetry-collector-contrib/extension/observer"
+)
+
+const (
+	otelHints            = "io.opentelemetry.discovery"
+	otelMetricsHints     = otelHints + ".metrics"
+	otelLogsHints        = otelHints + ".logs"
+	discoveryEnabledHint = "enabled"
+	scraperHint          = "scraper"
+	configHint           = "config"
+)
+
+// HintsTemplatesBuilder creates configuration templates from provided hints.
+type HintsTemplatesBuilder interface {
+	createScraperTemplateFromHints() ([]receiverTemplate, error)
+}
+
+// K8sHintsBuilder creates configurations from hints provided as Pod's annotations.
+type K8sHintsBuilder struct {
+	logger *zap.Logger
+	config K8sHintsConfig
+}
+
+// createReceiverTemplateFromHints creates a receiver configuration based on the provided hints.
+// Hints are extracted from Pod's annotations.
+// Scraper configurations are only created for Port Endpoints.
+// Log receiver configurations are only created for Pod Container Endpoints.
+func (builder *K8sHintsBuilder) createReceiverTemplateFromHints(env observer.EndpointEnv) (*receiverTemplate, error) {
+	var endpointType string
+	var podUID, podName, podNs string
+	var annotations map[string]string
+
+	endpointType = getStringEnv(env, "type")
+	if endpointType == "" {
+		return nil, fmt.Errorf("could not get endpoint type: %v", zap.Any("env", env))
+	}
+
+	if endpointType != string(observer.PortType) && endpointType != string(observer.PodContainerType) {
+		return nil, nil
+	}
+
+	builder.logger.Debug("handling hints for added endpoint", zap.Any("env", env))
+
+	if pod, ok := env["pod"]; ok {
+		endpointPod, ok := pod.(observer.EndpointEnv)
+		if !ok {
+			return nil, fmt.Errorf("could not extract endpoint's pod: %v", zap.Any("endpointPod", pod))
+		}
+		ann := endpointPod["annotations"]
+		if ann != nil {
+			annotations, ok = ann.(map[string]string)
+			if !ok {
+				return nil, fmt.Errorf("could not extract annotations: %v", zap.Any("annotations", ann))
+			}
+			if len(annotations) == 0 {
+				return nil, nil
+			}
+		}
+		podUID = endpointPod["uid"].(string)
+		podName = endpointPod["name"].(string)
+		podNs = endpointPod["namespace"].(string)
```

Comment:
Will these fields be always present? Wdyt adding some type-assertions or comment explaining why this fields will be always string and present in the k8s api?

Reply:
Good point. Actually I realize we can just use the observer's structs directly: 216998e34f95f0a8a4a975c2f4ab9290176d2222. 

However, that might require the changes in the `PodContainer` to happen in a separate PR 🤔 . (change from `container_name` to `name` etc)